### ISSUE TITLE
Black and white-listing RPCs

### ISSUE THREAD

**Issue by esso23** [Code-Sharp/WampSharp#259]
Hi Elad,

I have trouble figuring out how to black/white-list some of my RPCs. I have a callee with ~20 RPCs like:

```cs
    public interface ICallee
    {
        [WampProcedure("ping")]
        int Ping();

        // ....
    }
```

and I register the callee using:

```cs
    Task<IAsyncDisposable> RegisterCallee(object instance, ICalleeRegistrationInterceptor interceptor);
```

Now, I searched the docs:
https://wampsharp.net/release-notes/wampsharp-v1.2.6.41-beta-release-notes/ (there's only documentation for pub/sub)
https://wampsharp.net/wamp2/roles/callee/reflection-based-callee/ (not documented)
https://wampsharp.net/#advanced-profile-features (not documented)

The problem is that I have no control over the PublishOptions of the response to the call.
Am I missing something?

Thanks
Esso

**Comment by darkl** [comment]
Black and whitelisting for RPC is not supported. Last time I checked, the spec wasn't finalized and this wasn't supported in crossbar.io either.

**Comment by esso23** [comment]
Wow, I wasn't even considering that an option :-O
Anyway, is there a way to work around this? When I receive the call I could use
```cs 
WampInvocationContext.Current.InvocationDetails.CallerAuthenticationRole
```
and then just reject the call. But I really don't want to put this code in every single RPC - is there a way to intercept the call on callee proxy?

**Comment by darkl** [comment]
It sounds like you can write an Authorizer to deal with this problem.

**Comment by esso23** [comment]
I use authorizers to assign auth roles to clients, but the problem with this is that authorizers are implemented on router and I don't want to modify/reconfigure router everytime I add a new RPC to any of the clients. So using auth roles seems like a good idea to me, because I can black/white-list specific auth roles when I'm implementing the RPC.
It would make sense if I could specify this in RegisterOptions during RPC registration - then the router would know what to do when the RPC is called - since router is supposed to be as generic as possible in WAMP afaik.

**Comment by darkl** [comment]
>
>
>
The authorizer can generically check if a given authrole is allowed to call
a given method (using the CanCall method). It looks like what you wanted at
first.

Black/whitelisting as far as I remember is about whether the callee is
allowed to receive an invocation, and is not related to this scenario.

See also
https://github.com/wamp-proto/wamp-proto/issues/181
https://github.com/wamp-proto/wamp-proto/issues/184


Elad

**Comment by esso23** [comment]
Thanks a lot for posting those threads, I was thinking similarly as the OP of those threads.
Here is the most important comment:
https://github.com/wamp-proto/wamp-proto/issues/184#issuecomment-139463966
TL;DR for reference - all authorization logic should be on router, except:

> if the authorization depends on the app payload (args/kwargs)

Thanks again,
Esso
